File: android/feature/analytics/src/main/kotlin/xyz/klinik/feature/analytics/AnalyticsModel.kt
package xyz.klinik.feature.analytics

import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import xyz.klinik.network.AnalyticsApi
import xyz.klinik.network.ChannelReport
import xyz.klinik.network.Currency
import xyz.klinik.network.GeographyReport
import xyz.klinik.network.OccupancyReport
import xyz.klinik.network.ProcedureReport
import xyz.klinik.network.RevenueReport

/** The ranges a clinic actually asks for. */
enum class ReportRange {
    THIS_MONTH,
    LAST_THREE_MONTHS,
    THIS_YEAR,
    LAST_YEAR,
    ;

    val stringKey: String
        get() = "analytics.range." + when (this) {
            THIS_MONTH -> "thisMonth"
            LAST_THREE_MONTHS -> "lastThreeMonths"
            THIS_YEAR -> "thisYear"
            LAST_YEAR -> "lastYear"
        }

    /**
     * The window, as the API wants it.
     *
     * Ends at "now" rather than at the end of the period, so a report about
     * what has happened does not include tomorrow's bookings. The zone is the
     * device's: a clinic asking for "this month" means its own month.
     */
    fun bounds(now: Instant, zone: ZoneId): Pair<String, String> {
        val moment = ZonedDateTime.ofInstant(now, zone)

        val from = when (this) {
            THIS_MONTH -> moment.withDayOfMonth(1).toLocalDate().atStartOfDay(zone)
            LAST_THREE_MONTHS -> moment.minusMonths(3).toLocalDate().atStartOfDay(zone)
            THIS_YEAR -> moment.withDayOfYear(1).toLocalDate().atStartOfDay(zone)
            LAST_YEAR -> moment.withDayOfYear(1).minusYears(1).toLocalDate().atStartOfDay(zone)
        }

        val to = when (this) {
            // Up to the instant this year began, not to today: last year is a
            // closed period, and running it to now would mix two.
            LAST_YEAR -> moment.withDayOfYear(1).toLocalDate().atStartOfDay(zone).minusSeconds(1)
            else -> moment
        }

        return from.toInstant().toString() to to.toInstant().toString()
    }
}

sealed interface AnalyticsPhase {
    data object Loading : AnalyticsPhase
    data object Loaded : AnalyticsPhase

    /** Not a failure: an account without `analytics.read` simply has no panel. */
    data object NotPermitted : AnalyticsPhase
}

data class AnalyticsState(
    val phase: AnalyticsPhase = AnalyticsPhase.Loading,
    val range: ReportRange = ReportRange.THIS_YEAR,
    val currency: Currency = Currency.TRY,
    val procedures: ProcedureReport? = null,
    val geography: GeographyReport? = null,
    val revenue: RevenueReport? = null,
    val channels: ChannelReport? = null,
    val occupancy: OccupancyReport? = null,
)

/**
 * The clinic's numbers (spec M11).
 *
 * Five reports, fetched together because they are one screen and a panel that
 * fills in over five seconds is a panel somebody screenshots half-drawn.
 *
 * Each one is allowed to be missing on its own. Revenue and channels need
 * `finance.read`, which a doctor may not hold while still being allowed to see
 * how many operations they did — so one section answering 403 leaves the rest
 * of the panel alone rather than replacing it with an error.
 */
class AnalyticsModel(
    private val api: AnalyticsApi,
    private val clock: () -> Instant = Instant::now,
    private val zone: ZoneId = ZoneId.systemDefault(),
) {
    private val _state = MutableStateFlow(AnalyticsState())
    val state: StateFlow<AnalyticsState> = _state.asStateFlow()

    suspend fun choose(range: ReportRange) {
        _state.value = _state.value.copy(range = range)
        load()
    }

    suspend fun choose(currency: Currency) {
        _state.value = _state.value.copy(currency = currency)
        load()
    }

    suspend fun load() = coroutineScope {
        val current = _state.value

        _state.value = current.copy(phase = AnalyticsPhase.Loading)

        val (from, to) = current.range.bounds(clock(), zone)

        val procedures = async { optional { api.procedures(from, to) } }
        val geography = async { optional { api.geography(from, to) } }
        val revenue = async { optional { api.revenue(from, to, current.currency) } }
        val channels = async { optional { api.channels(from, to, current.currency) } }
        val occupancy = async { optional { api.occupancy(from, to) } }

        val loaded = listOf(procedures, geography, revenue, channels, occupancy).map { it.await() }

        _state.value = current.copy(
            // Nothing at all came back: this account cannot see the panel,
            // which is a different thing from an empty clinic and reads
            // differently.
            phase = if (loaded.all { it == null }) {
                AnalyticsPhase.NotPermitted
            } else {
                AnalyticsPhase.Loaded
            },
            procedures = loaded[0] as ProcedureReport?,
            geography = loaded[1] as GeographyReport?,
            revenue = loaded[2] as RevenueReport?,
            channels = loaded[3] as ChannelReport?,
            occupancy = loaded[4] as OccupancyReport?,
        )
    }

    private suspend fun <T> optional(work: suspend () -> T): T? =
        runCatching { work() }.getOrNull()
}

File: ios/Sources/KlinikAnalyticsFeature/AnalyticsModel.swift
import Foundation
import KlinikAPI
import KlinikCore

/// The ranges a clinic actually asks for, plus a way to say something else.
public enum ReportRange: String, Sendable, Equatable, CaseIterable, Identifiable {
    case thisMonth
    case lastThreeMonths
    case thisYear
    case lastYear

    public var id: String { rawValue }
    public var localizedName: String { L10n.string("analytics.range.\(rawValue)") }

    /// Half-open at the end so "this month" does not include tomorrow's
    /// bookings in a report about what has happened.
    public func bounds(now: Date = Date(), calendar: Calendar = .current) -> (from: Date, to: Date) {
        switch self {
        case .thisMonth:
            let start = calendar.date(
                from: calendar.dateComponents([.year, .month], from: now)
            ) ?? now

            return (start, now)

        case .lastThreeMonths:
            let start = calendar.date(byAdding: .month, value: -3, to: now) ?? now

            return (calendar.startOfDay(for: start), now)

        case .thisYear:
            let start = calendar.date(from: calendar.dateComponents([.year], from: now)) ?? now

            return (start, now)

        case .lastYear:
            let thisYear = calendar.date(from: calendar.dateComponents([.year], from: now)) ?? now
            let start = calendar.date(byAdding: .year, value: -1, to: thisYear) ?? now
            let end = calendar.date(byAdding: DateComponents(second: -1), to: thisYear) ?? now

            return (start, end)
        }
    }
}

public enum AnalyticsPhase: Sendable, Equatable {
    case loading
    case loaded
    /// Not a failure: an account without `analytics.read` simply has no panel.
    case notPermitted
    case failed(String)
}

public struct AnalyticsState: Sendable, Equatable {
    public var phase: AnalyticsPhase = .loading
    public var range: ReportRange = .thisYear
    public var currency: Currency = .turkishLira

    public var procedures: ProcedureReport?
    public var geography: GeographyReport?
    public var revenue: RevenueReport?
    public var channels: ChannelReport?
    public var occupancy: OccupancyReport?

    public init() {}
}

/**
 * The clinic's numbers (spec M11).
 *
 * Five reports, fetched together because they are one screen and a panel that
 * fills in over five seconds is a panel somebody screenshots half-drawn.
 *
 * Each one is allowed to be missing on its own. The revenue and channel
 * reports need `finance.read`, which a doctor may not hold while still being
 * allowed to see how many operations they did — so a 403 on one section leaves
 * the rest of the panel alone rather than replacing it with an error.
 */
@MainActor
public final class AnalyticsModel {
    private let api: AnalyticsAPI
    private var state = AnalyticsState()

    public init(api: AnalyticsAPI) {
        self.api = api
    }

    public func currentState() -> AnalyticsState { state }

    public func choose(range: ReportRange) async {
        state.range = range
        await load()
    }

    public func choose(currency: Currency) async {
        state.currency = currency
        await load()
    }

    public func load(now: Date = Date()) async {
        let bounds = state.range.bounds(now: now)

        async let procedures = attempt {
            try await self.api.procedures(from: bounds.from, to: bounds.to)
        }
        async let geography = attempt {
            try await self.api.geography(from: bounds.from, to: bounds.to)
        }
        async let revenue = attempt {
            try await self.api.revenue(from: bounds.from, to: bounds.to, currency: self.state.currency)
        }
        async let channels = attempt {
            try await self.api.channels(from: bounds.from, to: bounds.to, currency: self.state.currency)
        }
        async let occupancy = attempt {
            try await self.api.occupancy(from: bounds.from, to: bounds.to)
        }

        let loaded = await (procedures, geography, revenue, channels, occupancy)

        state.procedures = loaded.0.value
        state.geography = loaded.1.value
        state.revenue = loaded.2.value
        state.channels = loaded.3.value
        state.occupancy = loaded.4.value

        let errors = [
            loaded.0.error, loaded.1.error, loaded.2.error, loaded.3.error, loaded.4.error,
        ].compactMap { $0 }

        // Nothing at all came back. That means this account cannot see the
        // panel only when the server said so: a clinic nobody can reach is not
        // a permission problem, and telling somebody it is sends them to ask
        // for access they already have.
        guard errors.count == 5 else {
            state.phase = .loaded
            return
        }

        switch ReadOutcome.of(errors) {
        case .refused: state.phase = .notPermitted
        case .failed(let message): state.phase = .failed(message)
        }
    }
}
